crates/ink/codegen/src/generator/dispatch.rs

```diff
@@ -435,16 +436,38 @@ impl Dispatch<'_> {
                         .unwrap_or_else(|error| ::core::panic!("{}", error))
                 }
 
-                ::ink::env::decode_input::<
-                        <#storage_ident as ::ink::reflect::ContractMessageDecoder>::Type>()
-                    .map_err(|_| ::ink::reflect::DispatchError::CouldNotReadInput)
-                    .and_then(|decoder| {
-                        <<#storage_ident as ::ink::reflect::ContractMessageDecoder>::Type
-                            as ::ink::reflect::ExecuteDispatchable>::execute_dispatchable(decoder)
-                    })
-                    .unwrap_or_else(|error| {
-                        ::core::panic!("dispatching ink! message failed: {}", error)
-                    })
+                let dispatchable = match ::ink::env::decode_input::<
+                        <#storage_ident as ::ink::reflect::ContractMessageDecoder>::Type>() {
+                        Ok(decoded_dispatchable) => {
+                            ::ink::env::debug_println!("Result from `decode_input` {:?}", &decoded_dispatchable);
+                            decoded_dispatchable
+                        },
+                        Err(_decoding_error) => {
+                        ::ink::env::debug_println!("Result from `decode_input` {:?}", &_decoding_error);
+
+                        // TODO: Here will pick out the `Dispatch` error we're interested in telling the
```

Comment:
I think the only error to raise here is the `CouldNotReadInput`, the rest are already handled by `execute_dispatchable` :point_down:

Reply:
Not quite. Within the call stack of `decode_input` we actually raise a bunch of `DispatchError`s, but they get [swallowed by `__ink_MessageDispatch`'s `scale::Decode` implementation](https://github.com/paritytech/ink/blob/beccff6409100e30a37144d4acfff2ac83b7bf29/crates/ink/codegen/src/generator/dispatch.rs#L813). We then map this to the more generic `CouldNotReadInput` error